target_chains/ethereum/contracts/contracts/executor/Executor.sol

```diff
@@ -4,64 +4,91 @@ pragma solidity ^0.8.0;
 import "../pyth/PythGovernanceInstructions.sol";
 import "../wormhole/interfaces/IWormhole.sol";
 import "./ExecutorErrors.sol";
+import "./ExecutorState.sol";
 
-contract Executor {
+abstract contract Executor is ExecutorState {
     using BytesLib for bytes;
 
-    // Magic is `PTGM` encoded as a 4 byte data: Pyth Governance Message
-    // TODO: it's annoying that we can't import this from PythGovernanceInstructions
-    uint32 constant MAGIC = 0x5054474d;
+    event ContractUpgraded(
+        address oldImplementation,
+        address newImplementation
+    );
 
-    PythGovernanceInstructions.GovernanceModule constant MODULE =
-        PythGovernanceInstructions.GovernanceModule.EvmExecutor;
+    // We have different actions here for potential future extensibility
+    enum ExecutorAction {
+        // TODO: add an instruction to change the governance data source.
+        Execute, // 0
+        UpgradeContract // 1
+    }
 
-    // Instruction indicating that the executor contract on
-    // targetChainId at executorAddress should call the contract at callAddress
-    // with the provided callData
     struct GovernanceInstruction {
         PythGovernanceInstructions.GovernanceModule module;
         ExecutorAction action;
         uint16 targetChainId;
+        bytes payload;
+    }
+
+    struct UpgradeContractPayload {
+        address newImplementation;
+    }
```

Comment:
do we really need other instruction type? I think the contract could be its own owner, and then you could send the normal execute instruction telling the executor to call its own upgrade method.